### PR TITLE
Python debugger source & display command

### PR DESCRIPTION
**Pull request by DanielBallaSZTE**
Reworked source(src) command to display a given range of the code, defaults to 3.
Added display command which does the same as the source, except it displays the source code everytime a breakpoint is hit.

JerryScript-DCO-1.0-Signed-off-by: Daniel Balla dballa@inf.u-szeged.hu

### CODE REVIEW

File: jerry-debugger/jerry-client-ws.py [hunk 1]
```diff
@@ -728,6 +759,35 @@ def parse_source(debugger, data):
     else:
         logging.debug("No pending breakpoints")
 
+def printSource(self, line_num):
+    last_bp = self.debugger.last_breakpoint_hit
+    lines = last_bp.function.source.split('\n',last_bp.function.source.count('\n') - 1)
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: Even if the source is an empty string, the array should have at least one value. if "".split("\n") return with an array which contains an empty string: `[ "" ]` that is ok. If it returns with an empty array `[]` than we should manually add a "" to this array.

And one more question: why do we need a second argument to split?

File: tests/debugger/do_src.expected [hunk 2]
```diff
@@ -9,14 +9,16 @@ Stopped at tests/debugger/do_src.js:20
 (jerry-debugger) s
 Stopped at <unknown>:1
 (jerry-debugger) src
-f = function f() {
+
+1   > f = function f() {
```
Comment (zherczeg) [inline tests/debugger/do_src.expected]: Why there is an empty line before this? We also need tests for src with argument and the new display commands.

File: jerry-debugger/jerry-client-ws.py [hunk 3]
```diff
@@ -117,7 +117,7 @@ class JerryFunction(object):
     def __init__(self, is_func, byte_code_cp, source, source_name, line, column, name, lines, offsets):
         self.is_func = is_func
         self.byte_code_cp = byte_code_cp
-        self.source = source
+        self.source = source.split('\n')
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: Is `\n` enough? Perhaps `re.split("\r\n|[\r\n]", source)` would be better.

File: jerry-debugger/jerry-client-ws.py [hunk 4]
```diff
@@ -134,6 +134,9 @@ def __init__(self, is_func, byte_code_cp, source, source_name, line, column, nam
             self.lines[line] = breakpoint
             self.offsets[offset] = breakpoint
 
+        for i in range(len(self.source)):
+            self.source[i] = "%d%4s %s" % (i + 1, " ", self.source[i])
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: We don't need this preparation :) The line index is available at use.

File: jerry-debugger/jerry-client-ws.py [hunk 5]
```diff
@@ -728,6 +759,24 @@ def parse_source(debugger, data):
     else:
         logging.debug("No pending breakpoints")
 
+def printSource(self, line_num):
+    last_bp = self.debugger.last_breakpoint_hit
+    lines = last_bp.function.source
+    if last_bp:
+        print("Source: %s" % (last_bp.function.source_name))
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: please check whether `last_bp.function.source_name` is available. There is no point printing empty lines.

File: jerry-debugger/jerry-client-ws.py [hunk 6]
```diff
@@ -728,6 +759,24 @@ def parse_source(debugger, data):
...
+            end = len(last_bp.function.source) - 1
+        else:
+            start = max(last_bp.line - line_num-1, 0)
+            end = min(last_bp.line + line_num, len(last_bp.function.source)-1)
+
+        for i in range(start, end):
+            if i == last_bp.line - 1:
+                print("%s" % lines[i][:3], ">", lines[i][6:])
+            else:
+                print(lines[i])
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: This is too complex. Just remove the for above, and format the line here.

**Comment by DanielBallaSZTE** [comment]
Added those issued changes.

File: tests/debugger/do_display.cmd [hunk 7]
```diff
@@ -0,0 +1,13 @@
+b a
+b b
+b c
+b d
+display
```
Comment (zherczeg) [inline tests/debugger/do_display.cmd]: I think display without argument is pointless, I would throw an error.

File: jerry-debugger/jerry-client-ws.py [hunk 8]
```diff
@@ -312,11 +312,38 @@ def do_backtrace(self, args):
...
+
+    do_source = do_src
+
+    def do_display(self, args):
+        """ Toggle source code display after breakpoints """
+        if args:
+            try:
+                line_num = int(args)
+                if line_num < 0:
+                    print("Error: Positive integer number expected")
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: Non-negative integer number expected (0 disables this command).

File: jerry-debugger/jerry-client-ws.py [hunk 9]
```diff
@@ -728,6 +757,27 @@ def parse_source(debugger, data):
     else:
         logging.debug("No pending breakpoints")
 
+def printSource(self, line_num):
+    last_bp = self.debugger.last_breakpoint_hit
+    lines = last_bp.function.source
+    if last_bp:
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: you are using last_bp before this check. Please move `lines = last_bp.function.source` after this check.

File: tests/debugger/do_display.expected [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+(jerry-debugger) display
+Error: Display should be called with an argument
+(jerry-debugger) c
+Stopped at breakpoint:1 tests/debugger/do_display.js:15 (in a() at line:15, col:1)
+(jerry-debugger) display 2
+(jerry-debugger) c
+Stopped at breakpoint:2 tests/debugger/do_display.js:16 (in b() at line:16, col:1)
+Source: tests/debugger/do_display.js
+15   function a() { print("hi"); }
+16  >function b() { print("welcome"); }
```
Comment (zherczeg) [inline tests/debugger/do_display.expected]: I would put a space after >

File: jerry-debugger/jerry-client-ws.py (hunk 9 above)
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: Btw we could do an early return here:

    if not last_bp:
      return

This way we don't need to indent the lines below.

File: jerry-debugger/jerry-client-ws.py [hunk 11]
```diff
@@ -728,6 +757,27 @@ def parse_source(debugger, data):
...
+            start = 0
+            end = len(last_bp.function.source) - 1
+        else:
+            start = max(last_bp.line - line_num, 0)
+            end = min(last_bp.line + line_num-1, len(last_bp.function.source)-1)
+
+        for i in range(start, end):
+            spaces = (4-len(str(i+1)))*" "
+            if i == last_bp.line - 1:
+                print("%d%s>%s" % (i + 1, spaces, lines[i]))
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: Space after >

Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: What about:

    if i == last_bp.line - 1:
      string = " > "
    else
      string = "   "
    print("%4d%s%s" % (i + 1, string, lines[i]))

This way you don't need spaces

File: jerry-debugger/jerry-client-ws.py [hunk 12]
```diff
@@ -728,6 +757,27 @@ def parse_source(debugger, data):
     else:
         logging.debug("No pending breakpoints")
 
+def printSource(self, line_num):
```
Comment (galpeter) [inline jerry-debugger/jerry-client-ws.py]: Naming style: not camel case but underscores. Thus the name should be: `print_source`.
Additionally, we should not accept self (as we don't know what is it) but instead the debugger itself as the first argument.

File: jerry-debugger/jerry-client-ws.py [hunk 13]
```diff
@@ -312,11 +312,38 @@ def do_backtrace(self, args):
...
-        if self.disable_args(args):
+        if args:
+            try:
+                line_num = int(args)
+                if line_num < 0:
+                    print("Error: Positive integer number expected")
+                    return
+            except ValueError as val_errno:
+                print("Error: Positive integer number expected, %s" % (val_errno))
+                return
```
Comment (galpeter) [inline jerry-debugger/jerry-client-ws.py]: Maybe we could extract the convert logic to a helper method, so then we would not need to duplicate the code here and in the next method.

File: jerry-debugger/jerry-client-ws.py [hunk 14]
```diff
@@ -729,6 +737,46 @@ def parse_source(debugger, data):
...
+    elif is_src_cmd:
+        return debugger.default_viewrange
+    else:
+        print("Display should be called with an argument")
+        return -1
+
+
+def print_source(debugger, line_num):
+    last_bp = debugger.last_breakpoint_hit
+    if last_bp:
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: One of my comment is missing:

    if not last_bp:
      return

And remove the indentation form the code after.

File: tests/debugger/do_src.expected [hunk 15]
```diff
@@ -8,15 +8,13 @@ Stopped at breakpoint:1 tests/debugger/do_src.js:16 (in f() at line:15, col:1)
 Stopped at tests/debugger/do_src.js:20
 (jerry-debugger) s
 Stopped at <unknown>:1
-(jerry-debugger) src
-f = function f() {
-print('F2') }
+(jerry-debugger) source 0
```
Comment (zherczeg) [inline tests/debugger/do_src.expected]: Is `source 0` prints the current line??? Isn't `source 1` should do this?

File: jerry-debugger/jerry-client-ws.py [hunk 16]
```diff
@@ -729,6 +737,48 @@ def parse_source(debugger, data):
...
+        end = len(last_bp.function.source) - 1
+    else:
+        start = max(last_bp.line - line_num, 0)
+        end = min(last_bp.line + line_num-1, len(last_bp.function.source)-1)
+
+    for i in range(start, end):
+        if i == last_bp.line - 1:
+            print("%4d > %s" % (i + 1, lines[i]))
+        else:
+            print("%4d%3s%s" % (i + 1, " ",lines[i]))
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: print("%4d   %s" % (i + 1, lines[i]))

Is good enough :)

File: jerry-debugger/jerry-client-ws.py [hunk 17]
```diff
@@ -729,6 +737,48 @@ def parse_source(debugger, data):
         logging.debug("No pending breakpoints")
 
 
+def src_check_args(debugger, is_src_cmd, args):
```
Comment (galpeter) [inline jerry-debugger/jerry-client-ws.py]: We had a bit of miscommunication, sorry about that. What I meant originally that the try-except block could be extracted to it's own method not the whole thing.

File: jerry-debugger/jerry-client-ws.py [hunk 18]
```diff
@@ -729,6 +737,48 @@ def parse_source(debugger, data):
...
+    if not last_bp:
+        return
+
+    lines = last_bp.function.source
+    if last_bp.function.source_name:
+        print("Source: %s" % (last_bp.function.source_name))
+
+    if line_num == 0:
+        start = 0
+        end = len(last_bp.function.source) - 1
```
Comment (galpeter) [inline jerry-debugger/jerry-client-ws.py]: Do we really need to subtract 1? the range method below will generate numbers in the `[start, end)` range, that is the `end` number will not be reached

Comment (DanielBallaSZTE) [inline jerry-debugger/jerry-client-ws.py]: Yes we do, if it is not added there would be an empty line printed at the end of the source command.

Comment (galpeter) [inline jerry-debugger/jerry-client-ws.py]: I see. ok.

**Comment by galpeter** [comment]
I don't have any other comments, lgtm

**Pull request merged**